addons/website_mail/controllers/main.py

```diff
@@ -47,111 +46,90 @@ def _message_post_helper(res_model='', res_id=None, message='', token='', token_
 
         The rest of the kwargs are passed on to message_post()
     """
-    res = request.env[res_model].browse(res_id)
-    author_id = request.env.user.partner_id.id
-    if token and res and token == getattr(res.sudo(), token_field, None):
-        res = res.sudo()
-        if request.env.user == request.env['ir.model.data'].xmlid_to_object('base.public_user'):
-            author_id = (res.partner_id and res.partner_id.id) if hasattr(res, 'partner_id') else author_id
+    record = request.env[res_model].browse(res_id)
+    author_id = request.env.user.partner_id.id if request.env.user.partner_id else False
+    if token and record and token == getattr(record.sudo(), token_field, None):
+        record = record.sudo()
+        if request.env.user == request.env.ref('base.public_user'):
+            author_id = record.partner_id.id if hasattr(record, 'partner_id') else author_id
         else:
-            author_id = request.env.user.partner_id and request.env.user.partner_id.id
             if not author_id:
                 raise NotFound()
     elif sha_in:
         timestamp = int(sha_time)
-        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
-        shasign = sha1('%s%s%s%s' % (res_model, res_id, secret, timestamp))
+        secret_sudo = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        shasign = sha1('%s%s%s%s' % (res_model, res_id, secret_sudo, timestamp))
         if sha_in == shasign.hexdigest() and int(time()) < timestamp + 3600 * 24:
-            res = res.sudo()
+            record = record.sudo()
         else:
             raise NotFound()
-    return res.with_context({'mail_create_nosubscribe': nosubscribe}).message_post(body=message,
-                                                                                   message_type=kw.pop('message_type', False) or "comment",
-                                                                                   subtype=kw.pop('subtype', False) or "mt_comment",
+    return record.with_context(mail_create_nosubscribe=nosubscribe).message_post(body=message,
```

Comment:
Could be good, but there is already 8 params for this method. My screen is not big enough to see the whole line ...